### PR TITLE
statistics: improve estimation for index equal condition

### PR DESCRIPTION
**Pull request by eurekaka**
### What problem does this PR solve?

Issue Number: close https://github.com/pingcap/tidb/issues/17364

Problem Summary:

Wrong plan is chosen when index equal condition contains value which is out of the range of the histogram.

### What is changed and how it works?

What's Changed:

Compute the NDV for the prefix columns of the index which are used for equal condition, and use it for row count guess.

How it Works:

It is more reasonable than hard-coded constant `outOfRangeBetweenRate`.

### Related changes

N/A

### Check List 

Tests 

- Unit test

Side effects

N/A

### Release note 

- improve row count estimation for index equal condition

### CODE REVIEW

File: planner/core/testdata/analyze_suite_out.json [hunk 1]
```diff
@@ -340,5 +340,20 @@
       "UnionAll{IndexReader(Index(t4.b)[[-inf,2)])->IndexReader(Index(t4.b)[[-inf,2)])}",
       "TableReader(Table(t4)->Sel([eq(test.t4.a, 1) le(test.t4.b, 2)]))"
     ]
+  },
+  {
+    "Name": "TestIndexEqualUnknown",
+    "Cases": [
+      [
+        "IndexReader_6 2.03 root  index:IndexRangeScan_5",
+        "└─IndexRangeScan_5 2.03 cop[tikv] table:t, index:PRIMARY(a, c, b) range:[7639902,7639902], keep order:false"
```
Comment (lzmhhh123) [inline planner/core/testdata/analyze_suite_out.json:349]: We'd better also record the sql in results.

File: statistics/table.go [hunk 2]
```diff
@@ -406,7 +407,18 @@ func (coll *HistColl) getEqualCondSelectivity(idx *Index, bytes []byte, coverAll
...
 		}
-		// for range queries
+		// The equal condition only uses prefix columns of the index.
+		colIDs := coll.Idx2ColumnIDs[idx.ID]
+		var ndv int64
+		for i, colID := range colIDs {
+			if i >= usedColsLen {
+				break
+			}
+			ndv = mathutil.MaxInt64(ndv, coll.Columns[colID].NDV)
```
Comment (lzmhhh123) [inline statistics/table.go:417]: Why not use the `col1.ndv * col2.ndv * ...` instead of max of them?

Comment (eurekaka) [reply statistics/table.go:417]: If the 2 rows are like (1,1),(2,2), the NDV if 2. Also, using `max` is consistent with `getCardinality`.

**Pull request merged**